Alternate ScanAssemblyAndDependencies implementation

bording opened the pull request:
@timbussmann Here's my attempt at "cleaning up" the `ScanAssemblyAndDependencies` implementation.


While testing it, I noticed that the circular test wasn't actually able to load both assemblies, so I tweaked it a bit.

timbussmann commented on src/NServiceBus.Core.Tests/AssemblyScanner/When_scanning_assemblies_with_circular_dependencies.cs:
```diff
@@ -14,10 +13,15 @@ public void ReferencesNServiceBus_circular()
 
-            // Put ClassLibraryB.dll in CurrentDomain.BaseDirectory so it resolves correctly
-            var tempClassLibB = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "ClassLibraryB.dll");
+            // Put assemblies in TestContext.CurrentContext.TestDirectory so they resolve correctly
```
I wonder why're not pointing the assembly scanner directly to the circular subfolder instead?

timbussmann commented on src/NServiceBus.Core/Hosting/Helpers/AssemblyScanner.cs:
```diff
@@ -116,7 +116,7 @@ static List<Type> GetHandlerMessageTypes(List<Type> list)
         {
             assembly = null;
-            if (!IsIncluded(Path.GetFileNameWithoutExtension(assemblyPath)))
+            if (IsExcluded(Path.GetFileNameWithoutExtension(assemblyPath)))
```
👍 was thinking about this rename as well, must have forgotten about it

timbussmann commented on src/NServiceBus.Core/Hosting/Helpers/AssemblyScanner.cs:
```diff
@@ -151,84 +151,57 @@ bool TryLoadScannableAssembly(string assemblyPath, AssemblyScannerResults result
             {
                 processed[assembly.FullName] = true;
-                AddTypesToResult(assembly, results);
-                return;
```
this would cause the nsb.core references to be scanned although that's not needed?

timbussmann commented on src/NServiceBus.Core/Hosting/Helpers/AssemblyScanner.cs:
removing this part, would cause quite a few more Assembly.Load calls, wouldn't it?

timbussmann commented on src/NServiceBus.Core/Hosting/Helpers/AssemblyScanner.cs:
I think it would be correct to make the ShouldScanAssembly check before loading references

bording commented:
@timbussmann Take a look. I addressed your feedback, was able to create a failing test for the "transitive reference in a nested folder" issue and fixed it. With that working, I was able to simplify things even more.

timbussmann commented on src/NServiceBus.Core/Hosting/Helpers/AssemblyScanner.cs:
```diff
@@ -54,31 +54,46 @@ internal AssemblyScanner(Assembly assemblyToScan)
-                ScanAssemblyAndDependencies(assemblyToScan, processed, results);
+                if (ScanAssembly(assemblyToScan, processed))
+                {
+                    AddTypesToResult(assemblyToScan, results);
```
so in this case we would no longer include types from a referenced assembly of the target assembly, but I guess that's ok. The only usage only passes core and wants to receive core's types, so there shouldn't be an issue
